### ISSUE TITLE
Killing the dired buffer leaves preview window open

### ISSUE THREAD

**Issue by artelse** [protesilaos/dired-preview#22]
When I kill the dired buffer `C-x k`, it leaves the preview window open. The preview buffers remain as well. When I switch to another buffer from dired, sometimes the preview window remains open.
Let me know if you need more details.

**Comment by protesilaos** [comment]
> From: artelse ***@***.***>
> Date: Wed, 11 Dec 2024 06:35:13 -0800
>
> When I kill the dired buffer `C-x k`, it leaves the preview window
> open. The preview buffers remain as well. When I switch to another
> buffer from dired, sometimes the preview window remains open. Let me
> know if you need more details.

This is a regression caused by the recent changes I made. I will have
time to review everything on Saturday.

More to follow!

-- 
Protesilaos Stavrou
https://protesilaos.com

**Comment by protesilaos** [comment]
I made some changes today which hopefully fix this and the other issue you had told me about. If you can test the latest development version, please let me know if it works for you.

**Comment by artelse** [comment]
Improvements overall; when I preview buffers, the number of buffers stay within the max set. However, when I switch to another non-dired buffer, it deletes [p] buffers but leaves two intact. When switching backward and forward between, somethings there are three [p] buffer. Anything you want me to test further?

**Comment by protesilaos** [comment]
Thank you! It still seems buggy to me how it kills the preview buffers. I made another change earlier and continue to monitor the situation. It is tricky to debug though.

**Comment by protesilaos** [comment]
> when switching backward and forward between, somethings there are three [p] buffer. 

Is this happening with the latest commit? I know this was happening initially, but I thought I fixed it. If not, can you give me an idea of how the buffer names look? Maybe there is a pattern there that is not accounted for.

**Comment by artelse** [comment]
Yes, run the latest, checked against the latest commit and actual `dired-preview.el` in `elpa`.

I can observe when switching to another buffer, it leaves the last three preview buffers open. Going back and forth, I also see the actual selected preview is removed when switched to another buffer, leaving the two previous ones. Hope I explain this good enough.

I don't think a buffer pattern is involved here as it happens with a variety of different buffers.

Then occasionally I get this error:
```
Error in window-state-change-hook (dired-preview--close-previews-outside-dired): (error "Attempt to delete main window of frame #<frame *Messages* - GNU Emacs at lae 0x6339f1a70220>")
```

Please let me know how I can help further.

**Comment by protesilaos** [comment]
We had a lot of subtle bugs related to the state of the previews and how we were killing buffers. I made those of changes these past few days. I think now we are good. If you have time to test the package again, please let me know how it goes.

**Comment by artelse** [comment]
Definitely an improvement; no left-over preview buffers when switching to another buffer. However, when I kill the dired buffer and launch a new one, I get this error, preventing any more previews:
```
Error running timer ‘dired-preview-display-file’: (error "Attempt to delete main window of frame #<frame faxitron - GNU Emacs at lae 0x5e174c10f220>") [6 times]
Error in window-state-change-hook (dired-preview--close-previews-outside-dired): (error "Attempt to delete main window of frame #<frame *Messages* - GNU Emacs at lae 0x5e174c10f220>")
```
Testing again, it shows one preview window and then throws the error and no more previews.

**Comment by protesilaos** [comment]
Thank you! I just made some changes. I think they fix the issue, though I will only be able to test this tomorrow as I will run out of electricity soon.

**Comment by artelse** [comment]
I don't see changes, it behaves the same. So double checked again to see if I run the latest, and indeed I do.

**Comment by protesilaos** [comment]
I have made several changes since then, which I believe address the underlying problem. I will publish a new version today and see how that goes as more users experience the new changes. Thank you!
